doc/design/session.md

```diff
@@ -0,0 +1,72 @@
+# Design Doc: Session
+
+## Abstract
+
+This design doc proposes to have an object called *Session* which
+encapsulates the environment in which the computation graph is
+executed.
+
+The session is able to distinguish running a graph locally or
+remotely, using CPU only or using one or more GPUs. Different sessions
+have different runtime environments such as [scopes](./scope.md) and
+device contexts.
+
+
+## Background
+
+A computation graph runs in an environment which contains states such
+as the scope and device contexts. The current design has an implicit
+global session on which `paddle.eval()` is executed.
+
+Since the user is not able to explicitly switch between runtime
+environments such as the scope and the device contexts, the user
+cannot run a topology in two independent environments. For example, in
+reinforcement learning, the user may want to have a stale model for
+inference and a fresh model for training, and only replace the stale
+model with the fresh model periodically. Also, we have no concept that
+can encapsulate a remote environment that could execute a computation
+graph.
+
+We need a session concept to address above issues.
+
+## Session
+
+A session is an object that owns all runtime states such as scope,
```

Comment:
Should define what a session owns exactly, like:

- The "graph" to run locally or remotely.
- Exactly ***one*** scope, containing all tensor variables for all the subnets.
- Settings and hyper-parameters.

Reply:
Thanks! Updated, please review.